var/spack/repos/builtin/packages/kokkos/package.py

```diff
@@ -247,13 +252,12 @@ def cmake_args(self):
             if var in self.spec:
                 options.append("-D%s_DIR=%s" % (tpl, spec[tpl].prefix))
 
-        # we do not need the compiler wrapper from Spack
-        # set the compiler explicitly (may be Spack wrapper or nvcc-wrapper)
-        try:
+        if '+rocm' in self.spec:
+            options.append('-DCMAKE_CXX_COMPILER=%s' %
```

Comment:
similar logic needs to go in `setup_dependent_package` so that downstream projects are still getting the correct `kokkos_cxx`

Reply:
Is this the current recommended way to use Kokkos? Seems like a pain :/ it would mean the device compiler is used for all host code as well, and the openmp implementation of the device compiler is used, so that would cause headaches in dependent projects that use openmp too.

Can Kokkos be configured to only use llvm-amdgpu's clang for device code?

The other issue is that setting CMAKE_CXX_COMPILER like this will not work well with spack, since it bypasses the compiler wrapper.

Reply:
> similar logic needs to go in setup_dependent_package so that downstream projects are still getting the correct kokkos_cxx

This may be more complicated than just altering `setup_dependent_package`: you can set environment variables with this method but generally cannot otherwise affect the build. If dependents should all use the HIP compiler there are a couple possibilities:

* They all also include configuration related to enabling HIP/ROCm support (the Spack `cabana` package appears to be an example of this)
* All dependents require building with `hipcc` regardless of whether they have explicit configuration for HIP support (there do appear to be several packages which depend on `kokkos` and don't yet mention HIP)

I'd be surprised if the latter was true (e.g. - https://github.com/ROCm-Developer-Tools/HIP/blob/main/docs/markdown/hip_faq.md#on-hip-clang-can-i-link-hip-code-with-host-code-compiled-with-another-compiler-such-as-gcc-icc-or-clang-); in that case, this may require a hybrid approach like defining a compiler entry

Reply:
This is just following the model for using MPI compiler wrappers in downstream projects. Downstream projects can choose to use kokkos_cxx or not (just like they can choose to use mpicxx or not). Kokkos doesn't know what is device code and what isn't, particularly since all files are C++ files (.cc, .cpp, etc). 

There are other methods available now to only use device compilers for Kokkos-dependent files through some CMake magic.  Projects can either choose to use the kokkos_cxx compiler wrappers OR they can use compiler launcher tricks via CMake and use the regular C++ compiler.

I haven't ever seen the Spack compiler actually do anything useful for CMake builds. We've never seen issues just using the raw compiler for builds.

Reply:
> I haven't ever seen the Spack compiler actually do anything useful for CMake builds

Adding `-march=..` and  `-mtune=..` flags for micro arch optimization for host code (how does that interact with hipcc actually?) and fixing up rpaths to have uniform rpaths or runpaths. I think this can be helpful for cmake projects too.

Reply:
Ah. Good to know. Thanks. Kokkos chooses all the optimization flags internally for specific architectures, so that's less helpful.  The rpath could be useful. We haven't thought about that much.

Reply:
My assumption is that packages would only need to set CXX to hipcc if they include code that requires HIP support. Does depending on Kokkos automatically imply this (i.e. when `kokkos` is built with HIP support)?

My thought is that packages which need this may as well consult spec['hip']directly (rather than using `kokkos_cxx`). In the CUDA case I can see why this is done, since there is an `nvcc_wrapper` provided by `kokkos` (and `kokkos-nvcc-wrapper`), but I don't see an analog for HIP.

Reply:
It's really up to the project. In general, if Kokkos depends on HIP and your project uses Kokkos, you expect your project to depend on HIP.  The project could try to detect HIP dependence itself and choose the compiler. We're just providing a "performance portability" option where your package.py can be platform-independent as much as possible leaving the details to Kokkos.  Of course, you can choose not to use `kokkos_cxx` and choose the compiler yourself.